internal/c/parts/gui/gui.cpp

```diff
@@ -76,318 +62,267 @@ static void gui_free_tokens(char **tokens) {
 /// @param qbsTitle [OPTIONAL] Title of the notification
 /// @param qbsMessage [OPTIONAL] The message that will be displayed
 /// @param qbsIconType [OPTIONAL] Icon type ("info" "warning" "error")
-/// @param passed How many parameters were passed?
+/// @param passed Optional parameter mask
 void sub__guiNotifyPopup(qbs *qbsTitle, qbs *qbsMessage, qbs *qbsIconType, int32_t passed) {
-    static qbs *aTitle = nullptr;
-    static qbs *aMessage = nullptr;
-    static qbs *aIconType = nullptr;
-
-    if (!aTitle)
-        aTitle = qbs_new(0, 0);
-
-    if (!aMessage)
-        aMessage = qbs_new(0, 0);
-
-    if (!aIconType)
-        aIconType = qbs_new(0, 0);
+    std::string aTitle;
+    std::string aMessage;
+    std::string aIconType;
 
     if (passed & 1)
-        qbs_set(aTitle, qbs_add(qbsTitle, qbs_new_txt_len("\0", 1)));
-    else
-        qbs_set(aTitle, qbs_new_txt_len("\0", 1));
+        aTitle.assign((const char *)qbsTitle->chr, qbsTitle->len);
```

Comment:
We might consider wrapping this up in a function just so you can't forget to pass the length. I would probably fall for that one :D Something like a `qbs_to_string(std::string *, qbs *)`. If we also had a `string_to_qbs` that produced a new `qbs *` from a `std::string` you might not even need `libqb.h` anymore, just having the `struct qbs` would be enough since you never access anything inside it.

Reply:
That's a kickass idea. We'll need to put these conversion functions somewhere then. I'm thinking `qbs.h` & `qbs.cpp` in `internal/c/libqb/`. We'll also have to tie this to `build.mk`. Let me know what you think.

Reply:
Re-read your main comment. XD Yeah. It's best to leave the conversion stuff for a later as it requires some planning. Also, now that I'm thinking about, we'll need to do and use the conversions carefully to avoid copying strings as much as possible to avoid performance hits.